pkg/controller/performanceprofile/performanceprofile_controller.go

```diff
@@ -140,54 +143,230 @@ func (r *ReconcilePerformanceProfile) Reconcile(request reconcile.Request) (reco
+		RTRepoURL:       *performanceProfile.Spec.RealTimeKernel.RepoURL,
+	}
+	mc, err := machineconfig.NewPerformance(components.AssetsDir, data)
+	if err != nil {
+		return err
+	}
+
+	if err := controllerutil.SetControllerReference(performanceProfile, mc, r.scheme); err != nil {
+		return err
+	}
+
+	if err := r.createOrUpdateMachineConfig(mc); err != nil {
+		return err
+	}
+
+	// deploy kubelet config
+	kc := kubeletconfig.NewPerformance(performanceProfile.Spec.CPU.Reserved)
+	if err := controllerutil.SetControllerReference(performanceProfile, kc, r.scheme); err != nil {
+		return err
+	}
+	if err := r.createOrUpdateKubeletConfig(kc); err != nil {
+		return err
+	}
+
+	// deploy feature gate
+	fg := featuregate.NewLatencySensitive()
+	if err := controllerutil.SetControllerReference(performanceProfile, fg, r.scheme); err != nil {
+		return err
+	}
+	if err := r.createOrUpdateFeatureGate(fg); err != nil {
+		return err
+	}
+
+	// deploy network latency tuned
+	networkLatencyTuned, err := tuned.NewNetworkLatency(components.AssetsDir)
+	if err != nil {
+		return err
+	}
+
+	if err := controllerutil.SetControllerReference(performanceProfile, networkLatencyTuned, r.scheme); err != nil {
+		return err
+	}
+
+	if err := r.createOrUpdateTuned(networkLatencyTuned); err != nil {
+		return err
+	}
+
+	// deploy real time kernel tuned
+	realTimeKernelTuned, err := tuned.NewWorkerRealTimeKernel(
+		components.AssetsDir,
+		performanceProfile.Spec.NodeSelector,
+		performanceProfile.Spec.CPU.Isolated,
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := controllerutil.SetControllerReference(performanceProfile, realTimeKernelTuned, r.scheme); err != nil {
+		return err
+	}
+
+	if err := r.createOrUpdateTuned(realTimeKernelTuned); err != nil {
+		return err
+	}
+
+	// unpause machine config pool
+	return r.pauseMachineConfigPool(mcp.Name, false)
+}
+
+func (r *ReconcilePerformanceProfile) pauseMachineConfigPool(mcpName string, pause bool) error {
+	// TODO: in case if we will change nodeSelector to the machineConfigPoolSelector under the API,
+	// we will need to fetch the name of the machine config pool on runtime
+	updatedMcp, err := r.getMachineConfigPool(mcpName)
+	if errors.IsNotFound(err) {
+		klog.Warning("failed to set pause, the machine config pool does not exist, probably it was deleted")
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	updatedMcp.Spec.Paused = pause
+	return r.client.Update(context.TODO(), updatedMcp)
+}
+
+func (r *ReconcilePerformanceProfile) verifyPerformanceProfileParameters(performanceProfile *performancev1alpha1.PerformanceProfile) error {
+	if performanceProfile.Spec.CPU == nil {
+		return fmt.Errorf("you should provide CPU section")
+	}
+
+	if performanceProfile.Spec.CPU.Isolated == nil {
+		return fmt.Errorf("you should provide isolated CPU set")
+	}
+
+	if performanceProfile.Spec.CPU.NonIsolated == nil {
+		return fmt.Errorf("you should provide non isolated CPU set")
```

Comment:
Why? I mean, why can't we just compute the non-isolated set by difference? Or do we want to do it in a future PR?

Reply:
@fromanirh welcome back:)
Ok, in general, we can have 3 different workflows under the host:
- OS workflows
- some user application workflows(not the k8s one)
- containers workflows

and for each of these workflows, we can have the different CPU sets (one of the CPU set can be a subset of the other, but we can not precisely calculate values).

For example:
We set `reserved` to 0-3, it means these CPU's will not be used for container workflows, so these CPU's can be used or for OS workflows or for other user application that requires CPU isolation, and now we should separate CPU's that will be used for OS workflows and another user application, so we can now have CPU's 0-1 for OS workflows(`nonIsolated`), that leaves us with additional variable `isolated` that we can set to 2-5, it means that we have CPU's 2-3 for another user application that requires isolated CPU's and CPU's 4-5 for container workloads that requires isolated CPU's.

I hope it was clear enough, give me know if you need some additional information.

BTW, I do not say that we can not optimize it somehow, but for sure it will be part of another PR.

Reply:
> and now we should separate CPU's that will be used for OS workflows and another user application, so we can now have CPU's 0-1 for OS workflows(nonIsolated), that leaves us with additional variable isolated that we can set to 2-5, it means that we have CPU's 2-3 for another user application that requires isolated CPU's and CPU's 4-5 for container workloads that requires isolated CPU's.

It seems like everything can be expressed with ```reserved``` and ```isolated```

example. 8 core machine with 1 reserved isolated core, 1 reserved non-isolated core, and 6 isolated cores for container workloads. 

```
reserved: "0-1"
isolated: "1-7"
```

If you wanted to leave a general cores for container workloads as well, you could do something like this.

```
reserved: "0-1"
isolated: "1-5"
```

That would be 1 isolated core for reserved, 1 non-isolated core for reserved, 4 isolated cores for container workloads, and 2 non-isolated cores for container workloads. 

then non-isolated would be assumed to be "0,6-7" 

would that optimization make sense?

Reply:
At first glance, it looks fine. I will give it another thought and will reflect it under the PR :)

Reply:
> If you wanted to leave a general cores for container workloads as well, you could do something like this.
> 
> ```
> reserved: "0-1"
> isolated: "1-5"
> ```
> 
> That would be 1 isolated core for reserved, 1 non-isolated core for reserved, 4 isolated cores for container workloads, and 2 non-isolated cores for container workloads.
> 
> then non-isolated would be assumed to be "0,6-7"

On the second though, non-isolated in our context means that these CPUs will hardly be used by the operating system, so it does not good to use them for container workloads, so you still need the non-isolated parameter to specify what part of 0,6-7 will be used for OS workflows.

Reply:
>  so you still need the non-isolated parameter to specify what part of 0,6-7 will be used for OS workflows.

reserved is 0-1. so 0 would be used for OS and 6-7 for containers.  What am i missing here?

Reply:
do we have any more thoughts here on removing non-isolated?

Reply:
your suggestion makes sense to me on a first look, but I'm not familiar enough with isolated / reserved CPUs to have a strong opinion here...

Reply:
ok so our assumption that nonIsolatedCpu's should be a subset of `reserved` CPU's and does not have an intersection with `isolated` Cpu's, I believe it will be correct in most cases, still, I feel unsure regarding this kind of calculations for all use cases
@fromanirh @vladikr Guys, do you have some thoughts?

@davidvossel Can we leave this discussion for another PR, to avoid unmerge of this one?

Reply:
My take: first, let's move this (important) discussion to another issue/PR to let this PR move forward.

Second: `isolated` and `reserved` *in the current implementation* must not overlap. So the question boils down if we actually need cpus which are neiter isolated or reserved, which form a shared pool of cores on which container workload can run.
After a bit more thought I agree that non-reserved, non-isolated cores is a worthy concept, but we should present it nicely.

I think the takeaway is this little topic deserves more attention and perhaps a better abstraction, which is another hint we should move this discussion to a separate ticket.